root: make X11 really optional on macOS

aphecetche opened the pull request:
This is an attempt to get Root compiled with OpenGL but without requiring X11 on macOS.

Note that this is my first Spack PR... Hopefully I've correctly read the [contribution guide]( https://spack.readthedocs.io/en/latest/contribution_guide.html) though. You'll let me know :wink: 

Note also that I have some other changes in the root recipe that I'd like to propose (in other PRs) to "activate" two more variants : arrow and monalisa. I assumed it would be better to stage the modifications, but I can of course propose a combined PR if preferred.

aphecetche committed:
root: make X11 really optional on macOS

tldahlgren commented on var/spack/repos/builtin/packages/apple-glu/package.py:
```diff
@@ -0,0 +1,56 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

aphecetche replied:
indeed, those packages have been sitting in my private repo for a while :wink:
now fixed.

tldahlgren commented on var/spack/repos/builtin/packages/apple-opengl/package.py:
```diff
@@ -0,0 +1,68 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

tldahlgren reviewed (changes requested):
Only chiming in on the copyright ATM.

HadrienG2 commented on var/spack/repos/builtin/packages/apple-glu/package.py:
```diff
@@ -0,0 +1,56 @@
+
+    version("4.0.0")
+
+    provides("glu")
```
Should probably be...

```suggestion
    provides("glu", when=sys.platform == "darwin")
```

...as in other package ?

aphecetche replied:
indeed.
I should also specify the version (glu@1.3) I guess. 
will change.

HadrienG2 commented on var/spack/repos/builtin/packages/apple-glu/package.py:
```diff
@@ -0,0 +1,56 @@
+                "Platforms/MacOSX.platform/Developer/SDKs/MacOSX",
+            ),
+        )
+        # env.set("SPACK_APPLE_GLU","1")
```
Commented code should go away unless it is meant to be uncommented.

aphecetche replied:
ok

HadrienG2 commented on var/spack/repos/builtin/packages/apple-opengl/package.py:
```diff
@@ -0,0 +1,68 @@
+                "Platforms/MacOSX.platform/Developer/SDKs/MacOSX",
+            ),
+        )
+        # env.set("SPACK_APPLE_OPENGL","1")
```
See above.

HadrienG2 commented on var/spack/repos/builtin/packages/apple-opengl/package.py:
```diff
@@ -0,0 +1,68 @@
+        )
+        # env.set("SPACK_APPLE_OPENGL","1")
+
+    @property
+    def headers(self):
+        return HeaderList(
+            (
+                "/Applications/Xcode.app/Contents/Developer/",
+                "Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/",
+                "Library/Frameworks/OpenGL.framework/Headers",
+            )
+        )
```
Doesn't AppleGlu need a variation of that bit as well ?

aphecetche replied:
I only tested this within the `root` recipe, which uses CMake's `find_package(OpenGL)` and CMake was happy as is. Probably because the `glu.h` is in fact colocated with the `gl.h` one ?
Anyway I can certainly add it, yes.

HadrienG2 commented on var/spack/repos/builtin/packages/root/package.py:
```diff
@@ -211,11 +211,12 @@ class Root(CMakePackage):
+    depends_on("ftgl@2.4.0:", when="+opengl")
+    depends_on("glew", when="+opengl")
+    depends_on("gl2ps", when="+opengl")
+    if sys.platform != "darwin":
+        depends_on("gl", when="+x+opengl")
+        depends_on("glu", when="+x+opengl")
```
To pick up AppleGlu and AppleOpenGL, you probably need some sort of `depends_on("gl", when="+opengl")` even on Darwin + same for glu ?

aphecetche replied:
your comment makes sense. Will have to cross-check carefully why it worked (or so I believed...)  without.

aphecetche replied:
actually I think I'm mixing things up here. Most probably the `apple-g`l and `apple-glu` are actually not needed at all for the `root` recipe as CMake `find_package(OpenGL)` is correctly finding the Apple headers and libraries.
Those are needed in other packages I have in my stack.
Will confirm this and if that's the case then this PR will be much simpler :wink:

HadrienG2 commented on var/spack/repos/builtin/packages/root/package.py:
```diff
@@ -428,8 +430,11 @@ def cmake_args(self):
 
         # Options related to ROOT's ability to download and build its own
         # dependencies. Per Spack convention, this should generally be avoided.
+
         options += [
-            define_from_variant("builtin_afterimage", "x"),
+            define_from_variant("builtin_afterimage", "x")
+            if "platform=darwin" not in self.spec
+            else define("builtin_afterimage", True),
```
I think separately computing an `afterimage_enabled = "+x" in self.spec or "platform=darwin" in self.spec` boolean then doing a single `define("builtin_afterimage", afterimage_enabled)` option would be clearer.

aphecetche replied:
ok

aphecetche commented:
while browsing the other gl related packages (e.g. mesa and mesa-glu) I realise I should probably rename the `apple-opengl` package to `apple-gl` instead ?

aphecetche commented:
actually I think I'm mixing things up here. Most probably the `apple-g`l and `apple-glu` are actually not needed at all for the `root` recipe as CMake `find_package(OpenGL)` is correctly finding the Apple headers and libraries.
Those are needed in other packages I have in my stack.
Will confirm this and if that's the case then this PR will be much simpler :wink:

aphecetche commented:
FYI I've implemented some changes and forced pushed.

I've changed `apple-opengl` to `apple-gl` and fixed the `C_INCLUDE_PATH` in both packages that was not a string but a tuple (my bad, was trying to pacify the black formatter which was complaining about too long lines and I added commas where there should not have...)

HadrienG2 commented on var/spack/repos/builtin/packages/root/package.py:
```diff
@@ -211,11 +211,15 @@ class Root(CMakePackage):
+    depends_on("ftgl@2.4.0:", when="+opengl")
+    depends_on("glew", when="+opengl")
+    depends_on("gl2ps", when="+opengl")
+    if sys.platform != "darwin":
+        depends_on("gl", when="+x+opengl")
+        depends_on("glu", when="+x+opengl")
+    else:
+        depends_on("gl", when="+opengl")
+        depends_on("glu", when="+opengl")
```
Given that `conflicts("+opengl", when="~x")` if `sys.platform != darwin`, I think this is equivalent to the simpler...

```suggestion
    depends_on("gl", when="+opengl")
    depends_on("glu", when="+opengl")
```

HadrienG2 commented on var/spack/repos/builtin/packages/root/package.py:
```diff
@@ -428,8 +433,11 @@ def cmake_args(self):
         # Options related to ROOT's ability to download and build its own
         # dependencies. Per Spack convention, this should generally be avoided.
+
+        afterimage_enabled = "+x" if "platform=darwin" not in self.spec else True
```
I think that should be...

```suggestion
        afterimage_enabled = ("+x" in self.spec) if "platform=darwin" not in self.spec else True
```

aphecetche committed:
Update var/spack/repos/builtin/packages/root/package.py

Co-authored-by: Hadrien G. <knights_of_ni@gmx.com>

aphecetche committed:
Update var/spack/repos/builtin/packages/root/package.py

Co-authored-by: Hadrien G. <knights_of_ni@gmx.com>

tldahlgren commented on var/spack/repos/builtin/packages/apple-gl/package.py:
```diff
@@ -0,0 +1,58 @@
+
+    version("4.1.0")
+
+    provides("gl@4.1", when=sys.platform == "darwin")
```
The `when` option is not needed here or in the `apple-glu`.

Conceptually, a package provides the interface/API (or not) typically for specific version(s) and or variant(s).  Then `conflicts` are used if it doesn't work or apply.

aphecetche replied:
fixed. thanks.

tldahlgren commented on var/spack/repos/builtin/packages/apple-glu/package.py:
```diff
@@ -0,0 +1,58 @@
+
+    version("1.3.0")
+
+    provides("glu@1.3", when=sys.platform == "darwin")
```
As mentioned above, this is not needed as the `conflicts()` address the support issue.

aphecetche committed:
remove when clauses in provides

aphecetche commented:
@tldahlgren I believe I've addressed the comments, but I don't exactly know how to clear the "1 change requested" status ...
